examples/images/cifar10/train_cifar10.py

```diff
@@ -81,7 +89,8 @@ def train(argv):
     dataloader = torch.utils.data.DataLoader(
         dataset,
         batch_size=FLAGS.batch_size,
-        shuffle=True,
+        sampler=DistributedSampler(dataset) if FLAGS.parallel else None,
+        shuffle=False if FLAGS.parallel else True,
```

Comment:
hum. I am rather unsure about this. where do you shuffle the data then?

Reply:
Fair point. I found this warning in the [PyTorch docs](https://pytorch.org/docs/stable/data.html): 
![image](https://github.com/user-attachments/assets/14b8109e-3d1b-4b8d-824e-2dfeee54b12a)

So in my current implementation, `train_sampler.set_epoch(epoch)` is missing, which I will add now.

Reply:
Perfect. Once you have finished your change, I will run the code myself. Once I get it working, I will merge the PR.

Final question, can you try to load and run the existing checkpoints? I just want to be sure that people can reproduce our results. Thx.

Reply:
Ok, I refactored the training loop to use `num_epochs` instead of `FLAGS.total_steps`, since `sampler.set_epoch(epoch)` uses an epoch count. However, I think we need to change more than this. The PyTorch warning I pasted above mentions that we need to use `sampler.set_epoch(epoch)` "**before** creating the `DataLoader` iterator", but right now, the data loader iterator is created **once** before the training loop:
```python
from utils_cifar import infiniteloop

datalooper = infiniteloop(dataloader)
```
The way I would change this is by having a training loop like this:
```python
with trange(num_epochs, dynamic_ncols=True) as epoch_pbar:
    for epoch in epoch_pbar:
        epoch_pbar.set_description(f"Epoch {epoch + 1}/{num_epochs}")
        if sampler is not None:
            sampler.set_epoch(epoch)
        for batch_idx, data in enumerate(dataloader): 
                # step += 1  # tricky
                optim.zero_grad()
                x1 = data.to(rank)  # old: `x1 = next(datalooper)`
                [...]
```
Is this fine by you? IMO, what is a bit tricky is to handle the `step` counter correctly (based on which checkpoints are saved and some samples during training are generated). In a distributed setup, we'll have several processes running in parallel, and thus, we would probably save checkpoints and images multiple times (once per process)? However, since the filenames do not reflect the process ID, one process would also overwrite the files of the other. What do you think?

Reply:
About your question: When you say "existing checkpoints", which ones do you mean? I had once run the training and generation of samples on one GPU and gotten an FID of `3.8` (which is only slightly worse than the `3.5` you report).